tests/bluetooth/tester/src/btp_has.c

```diff
@@ -0,0 +1,204 @@
+			if (index) {
+				err = bt_has_preset_unregister(index);
+				has_presets &= ~(1<<(index-1));
+			}
+			else {
+				break;
+			}
+		}
+	}
+	else {
+		err = bt_has_preset_unregister(cp->index);
+	}
+	return (err) ? BTP_STATUS_FAILED : BTP_STATUS_SUCCESS;
+}
+
+static int has_preset_selected(unsigned char index, bool sync)
+{
+	return BTP_STATUS_SUCCESS;
+}
+
+static const struct bt_has_preset_ops has_preset_ops = { has_preset_selected, NULL };
+
+static uint8_t has_add_preset(const void *cmd, uint16_t cmd_len,
+			   void *rsp, uint16_t *rsp_len)
+{
+	const struct btp_has_add_preset_cmd *cp = cmd;
+
+	memcpy(temp_name, cp->name, cp->length);
+	temp_name[cp->length] = '\0';
+	struct bt_has_preset_register_param preset_params = { cp->index, cp->props, temp_name, &has_preset_ops };
+	int err = bt_has_preset_register(&preset_params);
+	if (!err) {
+		has_presets |= 1<<(cp->index-1);
+	}
+	return (err) ? BTP_STATUS_FAILED : BTP_STATUS_SUCCESS;
+}
+
+static uint8_t has_set_properties(const void *cmd, uint16_t cmd_len,
+			   void *rsp, uint16_t *rsp_len)
+{
+	const struct btp_has_set_properties_cmd *cp = cmd;
+	int err = (cp->props & BT_HAS_PROP_AVAILABLE) ? bt_has_preset_available(cp->index) : bt_has_preset_unavailable(cp->index);
+	return (err) ? BTP_STATUS_FAILED : BTP_STATUS_SUCCESS;
+}
+
+static const struct btp_handler has_handlers[] = {
+	{
+		.opcode = BTP_HAS_READ_SUPPORTED_COMMANDS,
+		.index = BTP_INDEX_NONE,
+		.expect_len = 0,
+		.func = has_supported_commands
+	},
+	{
+		.opcode = BTP_HAS_SET_ACTIVE_INDEX,
+		.expect_len = sizeof(struct btp_has_set_active_index_cmd),
+		.func = has_set_active_index
+	},
+	{
+		.opcode = BTP_HAS_SET_PRESET_NAME,
+		.expect_len = BTP_HANDLER_LENGTH_VARIABLE,
+		.func = has_set_preset_name
+	},
+	{
+		.opcode = BTP_HAS_REMOVE_PRESET,
+		.expect_len = sizeof(struct btp_has_remove_preset_cmd),
+		.func = has_remove_preset
+	},
+	{
+		.opcode = BTP_HAS_ADD_PRESET,
+		.expect_len = BTP_HANDLER_LENGTH_VARIABLE,
+		.func = has_add_preset
+	},
+	{
+		.opcode = BTP_HAS_SET_PROPERTIES,
+		.expect_len = sizeof(struct btp_has_set_properties_cmd),
+		.func = has_set_properties
+	}
+};
+
+static const char *preset_name(uint8_t index)
+{
+	switch (index) {
+		case 0: return "PRESET_0";
+		case 1: return "PRESET_1";
+		case 2: return "PRESET_2";
+		case 3: return "PRESET_3";
+		case 4: return "PRESET_4";
+		case 5: return "PRESET_5";
+		case 6: return "PRESET_6";
+		case 7: return "PRESET_7";
+		default: return "PRESET_?";
+	}
+}
+
+uint8_t tester_init_has(void)
+{
+	tester_register_command_handlers(BTP_SERVICE_ID_HAS, has_handlers,
+					 ARRAY_SIZE(has_handlers));
+
+	struct bt_has_register_param params = { BT_HAS_HEARING_AID_TYPE_BINAURAL, false, true };
+	int err = bt_has_register(&params);
+	if (!err) {
+		for (uint8_t index = 1; index <= MIN(3, CONFIG_BT_HAS_PRESET_COUNT); index++) {
+			enum bt_has_properties properties = (index < 3) ? BT_HAS_PROP_WRITABLE | BT_HAS_PROP_AVAILABLE : BT_HAS_PROP_WRITABLE;
```

Comment:
Why the `MIN(3`?

Reply:
3 - is the number of presets being configured at initialization.  Perhaps a define would be more readable.